k2/csrc/fsa_utils.cu

```diff
@@ -291,9 +291,8 @@ class OpenFstStreamReader {
       line_is >> cost;
       if (!line_is.eof()) line_is >> std::ws;
     }
-    if (!line_is.eof() || line_is.fail() || src_state < 0 || dest_state < 0) {
+    if (!line_is.eof() || src_state < 0 || dest_state < 0) {
```

Comment:
I don't think it's safe to ignore line_is.fail() here.  E.g. above, it does:
`ExpectChar(line_is, ']');  // sets failbit if not..`
That check would be ineffective if we ignore failbit.   If this is failing we need to figure out exactly why, I think.

Reply:
> I don't think it's safe to ignore line_is.fail() here. E.g. above, it does:
ExpectChar(line_is, ']'); // sets failbit if not..

In that case, I recommend throwing as soon as we detect that `line_is.fail()` is `true`.
I just reverted the changes and change `if(!line_is.good())` to `if(line.fail())` and 
someone just tested it and the problem still exists.

Here is the error message:
[error.txt](https://github.com/k2-fsa/k2/files/9052985/error.txt)

Reply:
@danpovey 
> I don't think it's safe to ignore line_is.fail() here. E.g. above, it does:
ExpectChar(line_is, ']'); // sets failbit if not..

I have changed the semantics of `ExpectChar`: If the char is unexpected, we invoke
`K2_LOG(FATAL)` so that we don't need to use `line.is_fail()` later.

Reply:
Sorry, I don't think the change in ExpectChar is good, because the current error-checking code prints out the line that failed, and other information like src_state.  Inside ExpectChar, we can't print out that info.  If we have a failure in future, this change would make it hard to debug.
I had a look at the error message, which was:
```
k2::OpenFstStreamReader::ProcessLine(std::string&) Invalid line: 3	0	137054	0, eof=true, fail=true, src_state=3, dest_state=0
```
... this is very hard to understand.  I cannot find a code path that should cause these results.  What we could perhaps do is if there is a way to reproduce the error, add other commands to detect at what point failbit is set.  (or find someone who is willing to get this in gdb and step through it).  

Changing:
```
    if (line_is.fail())  {
```
to
```
    if (line_is.good())  {
```
should be safe, even if it is not necessary or does not help here.
I would rather not mask this error.  If something here is not behaving as it should, that indicates a deeper problem, and we should find the problem before we attempt to mask it.